File: py/src/verifiable_ai_sdk/core/provenance.py
"""Verify GitHub image build attestations with Sigstore's production trust root.

The caller supplies the expected build identity. These helpers do not discover
trusted publishers, approve deployment versions, or inspect running containers.
"""

from __future__ import annotations

import asyncio
import json
import re
from collections.abc import Sequence
from dataclasses import dataclass
from urllib.parse import parse_qsl, quote, urlencode, urlsplit

from cryptography import x509
from pydantic import ValidationError
from sigstore.errors import Error as SigstoreError
from sigstore.errors import VerificationError as SigstoreVerificationError
from sigstore.models import Bundle
from sigstore.verify import Verifier
from sigstore.verify.policy import Identity

from ..schemas import GitHubImageAttestationsSchema, SlsaStatementSchema
from ..types.provenance import ImageProvenancePolicy, VerifiedImageProvenance
from ..utils.errors import VerificationError, api_failure, verification_failure
from ..utils.fetch import fetch


_DIGEST_PATTERN = re.compile(r'sha256:[0-9a-fA-F]{64}')
_REPOSITORY_PATTERN = re.compile(r'[A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+')
_COMMIT_PATTERN = re.compile(r'[0-9a-fA-F]{40}')
_SOURCE_REPOSITORY_DIGEST_OID = x509.ObjectIdentifier('1.3.6.1.4.1.57264.1.13')
_GITHUB_WORKFLOW_SHA_OID = x509.ObjectIdentifier('1.3.6.1.4.1.57264.1.3')


async def fetch_image_provenance(
    repository: str, digest: str, github_token: str | None = None
) -> list[str]:
    """Fetch inline Sigstore bundles for an image digest from a GitHub repository.

    ``github_token`` is an optional GitHub token, not a Gateway API key. Only
    api.github.com is requested; returned bundle URLs are never followed.
    """

    if not _REPOSITORY_PATTERN.fullmatch(repository) or any(
        part in ('.', '..') for part in repository.split('/')
    ):
        raise api_failure(
            'api.invalid_input',
            {
                'field': 'repository',
                'reason': 'unsupported_value',
                'expected': 'owner/repo',
            },
        )
    if not _DIGEST_PATTERN.fullmatch(digest):
        raise api_failure(
            'api.invalid_input',
            {
                'field': 'digest',
                'reason': 'unsupported_value',
                'expected': 'sha256:<64 hex characters>',
            },
        )
    headers = {
        'Accept': 'application/vnd.github+json',
        'X-GitHub-Api-Version': '2022-11-28',
    }
    if github_token is not None:
        headers['Authorization'] = f'Bearer {github_token}'

    base_url = (
        f'https://api.github.com/repos/{repository}/attestations/'
        f'{quote(digest.lower(), safe=":")}?per_page=100'
    )
    url = base_url
    visited = {url}
    bundles: list[str] = []
    while True:
        try:
            response = await fetch(url, headers=headers)
        except Exception as error:
            raise api_failure(
                'api.transport_failed',
                {'resource': 'image_provenance', 'reason': 'request'},
                retryable=True,
                cause=error,
            ) from error
        if not response.ok:
            raise api_failure(
                'api.http_status',
                {'resource': 'image_provenance', 'status': response.status},
                retryable=response.status == 429 or response.status >= 500,
            )
        try:
            raw = json.loads(response.body)
        except (ValueError, UnicodeDecodeError) as error:
            raise api_failure(
                'api.invalid_json',
                {'resource': 'image_provenance'},
                cause=error,
            ) from error
        try:
            parsed = GitHubImageAttestationsSchema.model_validate(raw)
        except ValidationError as error:
            raise api_failure(
                'api.invalid_response',
                {
                    'resource': 'image_provenance',
                    'path': 'attestations',
                    'expected': 'GitHub attestations with inline bundles',
                    'actual': 'invalid response',
                },
                cause=error,
            ) from error
        bundles.extend(json.dumps(item.bundle) for item in parsed.attestations)
        link = ', '.join(
            value for name, value in response.headers.items() if name.lower() == 'link'
        )
        try:
            next_url = _next_attestation_url(link, base_url)
            if next_url is None:
                return bundles
            if next_url in visited:
                raise ValueError('repeated pagination cursor')
        except ValueError as error:
            raise api_failure(
                'api.invalid_response',
                {
                    'resource': 'image_provenance',
                    'path': 'Link',
                    'expected': 'a next link with a new before or after cursor',
                    'actual': 'invalid or repeated pagination cursor',
                },
                cause=error,
            ) from error
        visited.add(next_url)
        url = next_url


def _next_attestation_url(link: str, base_url: str) -> str | None:
    for entry in re.split(r',\s*(?=<)', link):
        relation = re.search(r';\s*rel\s*=\s*(?:"([^"]*)"|([^;\s]+))', entry)
        if relation is None or 'next' not in (relation[1] or relation[2] or '').split():
            continue
        target = re.match(r'\s*<([^>]*)>', entry)
        if target is None:
            raise ValueError('invalid pagination link')
        parsed_url = urlsplit(target[1])
        if not parsed_url.scheme or not parsed_url.netloc:
            raise ValueError('invalid pagination URL')
        cursors = [
            (name, value)
            for name, value in parse_qsl(
                parsed_url.query, keep_blank_values=True, errors='strict'
            )
            if name in ('before', 'after')
        ]
        if len(cursors) != 1 or not cursors[0][1]:
            raise ValueError('missing or ambiguous pagination cursor')
        # The link contributes only its cursor, never its origin or path.
        return f'{base_url}&{urlencode(cursors)}'
    return None


async def verify_image_provenance(
    bundles: Sequence[str], digest: str, policy: ImageProvenancePolicy
) -> VerifiedImageProvenance:
    """Accept any bundle that verifies the digest and caller's build policy.

    Sigstore authenticates the certificate, DSSE signature and transparency-log
    evidence. Only then do we interpret the in-toto statement and match its
    subject digest, source repository and certificate's source commit before
    applying the optional reviewed commit pin.
    """

    if not _DIGEST_PATTERN.fullmatch(digest):
        raise verification_failure(
            'input.invalid',
            {
                'field': 'digest',
                'reason': 'unsupported_value',
                'expected': 'sha256:<64 hex characters>',
            },
        )
    if not bundles:
        raise _provenance_failure(digest, ['no_attestations'])
    # Sigstore refreshes its TUF trust root synchronously. Keep this and the
    # subsequent certificate/signature checks off the caller's asyncio loop.
    return await asyncio.to_thread(_verify_bundles, bundles, digest.lower(), policy)


class _IdentityMismatch(SigstoreVerificationError):
    pass


class _StatementMismatch(Exception):
    def __init__(self, reason: str) -> None:
        self.reason = reason
        super().__init__(reason)


@dataclass
class _GitHubBuildIdentity:
    expected: ImageProvenancePolicy
    identity: str = ''
    ref: str = ''

    def verify(self, cert: x509.Certificate) -> None:
        prefix = (
            f'https://github.com/{self.expected.repository}/{self.expected.workflow}@'
        )
        try:
            san = cert.extensions.get_extension_for_class(x509.SubjectAlternativeName)
            identities = san.value.get_values_for_type(x509.UniformResourceIdentifier)
            for identity in identities:
                if not identity.startswith(prefix):
                    continue
                ref = identity[len(prefix) :]
                if not ref.startswith('refs/'):
                    continue
                if self.expected.ref is not None and ref != self.expected.ref:
                    continue
                Identity(identity=identity, issuer=self.expected.issuer).verify(cert)
                self.identity = identity
                self.ref = ref
                return
        except (x509.ExtensionNotFound, SigstoreVerificationError) as error:
            raise _IdentityMismatch(
                'GitHub build identity does not match policy'
            ) from error
        raise _IdentityMismatch('GitHub build identity does not match policy')


def _verify_bundles(
    bundles: Sequence[str], digest: str, policy: ImageProvenancePolicy
) -> VerifiedImageProvenance:
    try:
        verifier = Verifier.production()
    except Exception as error:
        raise _provenance_failure(digest, ['trust_root_unavailable'], error) from error

    reasons: list[str] = []
    last_error: BaseException | None = None
    for raw in bundles:
        identity = _GitHubBuildIdentity(policy)
        try:
            bundle = Bundle.from_json(raw)
            content_type, payload = verifier.verify_dsse(bundle, identity)
            if content_type != 'application/vnd.in-toto+json':
                raise _StatementMismatch('invalid_statement')
            try:
                statement = SlsaStatementSchema.model_validate_json(payload)
            except ValidationError as error:
                raise _StatementMismatch('invalid_statement') from error
            commit = _verify_statement(
                statement, digest, policy, identity.ref, bundle.signing_certificate
            )
            return VerifiedImageProvenance(
                digest=digest,
                repository=policy.repository,
                workflow=policy.workflow,
                ref=identity.ref,
                commit=commit,
                certificate_identity=identity.identity,
                issuer=policy.issuer,
                predicate_type=statement.predicate_type,
            )
        except _IdentityMismatch as error:
            reason = 'untrusted_identity'
            last_error = error
        except _StatementMismatch as error:
            reason = error.reason
            last_error = error
        except (SigstoreError, ValueError, TypeError) as error:
            reason = 'invalid_bundle'
            last_error = error
        if reason not in reasons:
            reasons.append(reason)
    raise _provenance_failure(digest, reasons, last_error)


def _verify_statement(
    statement: SlsaStatementSchema,
    digest: str,
    policy: ImageProvenancePolicy,
    ref: str,
    certificate: x509.Certificate,
) -> str:
    expected_hash = digest.removeprefix('sha256:')
    if not any(
        subject.digest.get('sha256', '').lower() == expected_hash
        for subject in statement.subject
    ):
        raise _StatementMismatch('digest_mismatch')

    expected_repository = f'https://github.com/{policy.repository}'
    commit: str | None = None
    if statement.predicate_type == 'https://slsa.dev/provenance/v1':
        definition = statement.predicate.build_definition
        if definition is None:
            raise _StatementMismatch('invalid_statement')
        workflow = definition.external_parameters.workflow
        if (
            workflow.repository.removesuffix('.git') != expected_repository
            or workflow.path != policy.workflow
            or workflow.ref != ref
        ):
            raise _StatementMismatch('source_mismatch')
        for dependency in definition.resolved_dependencies:
            if _source_matches(dependency.uri, expected_repository, ref):
                commit = dependency.digest.get('gitCommit')
                break
    else:
        invocation = statement.predicate.invocation
        if invocation is None:
            raise _StatementMismatch('invalid_statement')
        source = invocation.config_source
        if (
            not _source_matches(source.uri, expected_repository, ref)
            or source.entry_point != policy.workflow
        ):
            raise _StatementMismatch('source_mismatch')
        commit = source.digest.get('sha1')
    if commit is None:
        raise _StatementMismatch('source_mismatch')
    if not _COMMIT_PATTERN.fullmatch(commit):
        raise _StatementMismatch('source_mismatch')
    if commit.lower() != _certificate_source_commit(certificate):
        raise _StatementMismatch('source_mismatch')
    if policy.commit is not None and commit.lower() != policy.commit.lower():
        raise _StatementMismatch('commit_mismatch')
    return commit.lower()


def _certificate_source_commit(certificate: x509.Certificate) -> str:
    for oid, prefix in (
        (_SOURCE_REPOSITORY_DIGEST_OID, b'\x0c\x28'),
        (_GITHUB_WORKFLOW_SHA_OID, b''),
    ):
        try:
            extension = certificate.extensions.get_extension_for_oid(oid).value
        except x509.ExtensionNotFound:
            continue
        if not isinstance(extension, x509.UnrecognizedExtension):
            raise _StatementMismatch('source_mismatch')
        # The modern field is a DER UTF8String with a 40-byte SHA; the legacy
        # field is raw text. A present but malformed modern field cannot fall back.
        value = extension.value
        if not value.startswith(prefix):
            raise _StatementMismatch('source_mismatch')
        commit = value[len(prefix) :]
        if re.fullmatch(rb'[0-9a-fA-F]{40}', commit) is None:
            raise _StatementMismatch('source_mismatch')
        return commit.decode('ascii').lower()
    raise _StatementMismatch('source_mismatch')


def _source_matches(uri: str, repository: str, ref: str) -> bool:
    source, separator, source_ref = uri.removeprefix('git+').partition('@')
    return (
        bool(separator)
        and source.removesuffix('.git') == repository
        and source_ref == ref
    )


def _provenance_failure(
    digest: str, reasons: list[str], cause: BaseException | None = None
) -> VerificationError:
    return verification_failure(
        'provenance.image_verification_failed',
        {'digest': digest, 'reasons': reasons},
        retryable=reasons == ['trust_root_unavailable'],
        cause=cause,
    )

File: js/src/core/provenance.ts
import { Buffer } from 'buffer';
import * as v from 'valibot';
import {
  AnyOf,
  Identity,
  OIDCIssuer,
  OIDCIssuerV2,
  PolicyError,
  SigstoreVerifier,
  TrustedRootProvider,
} from '@freedomofpress/sigstore-browser';
import {
  GitHubImageAttestationsSchema,
  ImageProvenanceBundleSchema,
  ImageProvenanceStatementSchema,
} from '../schemas';
import type {
  FetchImageProvenanceParams,
  ImageProvenanceFailureReason,
  ImageProvenancePolicy,
  ImageProvenanceStatement,
  VerifiedImageProvenance,
  VerifyImageProvenanceParams,
} from '../types/provenance';
import { ApiError, VerificationError, inputError } from '../utils/errors';

const GITHUB_ISSUER = 'https://token.actions.githubusercontent.com';
// TUF authenticates root updates. In-memory caching works in both Node and
// browsers, without requiring IndexedDB or sharing application credentials.
const trustedRootProvider = new TrustedRootProvider({ disableCache: true });

/** Fetch all inline Sigstore bundles published for a repository's image digest. */
export async function fetchImageProvenance({
  repository,
  digest,
  githubToken,
}: FetchImageProvenanceParams): Promise<readonly string[]> {
  if (
    !/^[\w.-]+\/[\w.-]+$/.test(repository) ||
    repository.split('/').some((part) => part === '.' || part === '..')
  ) {
    throw invalidApiInput('repository', 'Expected a GitHub owner/repository');
  }
  if (!isSha256Digest(digest)) {
    throw invalidApiInput(
      'digest',
      'Expected sha256 followed by 64 hex digits',
    );
  }
  const repositoryPath = repository
    .split('/')
    .map(encodeURIComponent)
    .join('/');
  const headers: Record<string, string> = {
    Accept: 'application/vnd.github+json',
    'X-GitHub-Api-Version': '2022-11-28',
  };
  if (githubToken !== undefined) {
    headers.Authorization = `Bearer ${githubToken}`;
  }

  const bundles: string[] = [];
  for (let page = 1; ; page++) {
    const url = `https://api.github.com/repos/${repositoryPath}/attestations/${encodeURIComponent(digest.toLowerCase())}?per_page=100&page=${page}`;
    let response: Response;
    try {
      response = await fetch(url, { headers });
    } catch (cause) {
      throw new ApiError(
        {
          code: 'api.transport_failed',
          details: { resource: 'image_provenance', reason: 'request' },
          retryable: true,
        },
        { cause },
      );
    }
    if (!response.ok) {
      throw new ApiError({
        code: 'api.http_status',
        details: { resource: 'image_provenance', status: response.status },
        retryable: response.status === 429 || response.status >= 500,
      });
    }

    let raw: unknown;
    try {
      raw = await response.json();
    } catch (cause) {
      throw new ApiError(
        {
          code: 'api.invalid_json',
          details: { resource: 'image_provenance' },
        },
        { cause },
      );
    }
    const parsed = v.safeParse(GitHubImageAttestationsSchema, raw);
    if (!parsed.success) {
      throw new ApiError({
        code: 'api.invalid_response',
        details: {
          path: 'attestations',
          expected: 'An array of inline Sigstore bundles',
          actual: 'Invalid GitHub attestation response',
        },
      });
    }
    bundles.push(
      ...parsed.output.attestations.map(({ bundle }) => JSON.stringify(bundle)),
    );
    if (parsed.output.attestations.length < 100) {
      return bundles;
    }
  }
}

/**
 * Verify a GitHub Actions SLSA build proof for an image manifest digest.
 * One complete matching proof is sufficient. This checks the caller's build
 * policy; it does not prove which containers are running now.
 */
export async function verifyImageProvenance({
  bundles,
  digest,
  policy,
}: VerifyImageProvenanceParams): Promise<VerifiedImageProvenance> {
  if (!isSha256Digest(digest)) {
    throw inputError({ field: 'digest', reason: 'invalid_hex' });
  }
  const normalizedDigest = digest.toLowerCase();
  if (bundles.length === 0) {
    throw imageFailure({
      digest: normalizedDigest,
      reasons: ['no_attestations'],
    });
  }

  const verifier = new SigstoreVerifier();
  try {
    await verifier.loadSigstoreRootWithTUF(trustedRootProvider);
  } catch (cause) {
    throw imageFailure({
      digest: normalizedDigest,
      reasons: ['trust_root_unavailable'],
      cause,
    });
  }

  const reasons = new Set<ImageProvenanceFailureReason>();
  let lastCause: unknown;
  for (const bundleJson of bundles) {
    try {
      return await verifyBundle({
        verifier,
        bundleJson,
        digest: normalizedDigest,
        policy,
      });
    } catch (cause) {
      lastCause = cause;
      if (
        cause instanceof VerificationError &&
        cause.failure.code === 'provenance.image_verification_failed'
      ) {
        for (const reason of cause.failure.details.reasons) reasons.add(reason);
      } else {
        reasons.add(
          cause instanceof PolicyError
            ? 'untrusted_identity'
            : 'invalid_bundle',
        );
      }
    }
  }
  throw imageFailure({
    digest: normalizedDigest,
    reasons: [...reasons],
    cause: lastCause,
  });
}

type VerifyBundleParams = {
  verifier: SigstoreVerifier;
  bundleJson: string;
  digest: string;
  policy: ImageProvenancePolicy;
};

async function verifyBundle({
  verifier,
  bundleJson,
  digest,
  policy,
}: VerifyBundleParams): Promise<VerifiedImageProvenance> {
  const raw: unknown = JSON.parse(bundleJson);
  const parsed = v.safeParse(ImageProvenanceBundleSchema, raw);
  if (!parsed.success)
    throw imageFailure({ digest, reasons: ['invalid_bundle'] });

  const issuer = policy.issuer ?? GITHUB_ISSUER;
  const identityPrefix = `https://github.com/${policy.repository}/${policy.workflow}@`;
  let certificateIdentity = '';
  let ref = '';
  const verified = await verifier.verifyDsse(parsed.output, {
    verify(cert) {
      const identity = cert.extSubjectAltName?.uri ?? '';
      const candidateRef = identity.slice(identityPrefix.length);
      if (
        !identity.startsWith(identityPrefix) ||
        !candidateRef.startsWith('refs/') ||
        (policy.ref !== undefined && policy.ref !== candidateRef)
      ) {
        throw new PolicyError(
          'The certificate does not match the required repository, workflow and ref',
        );
      }
      new Identity({ identity }).verify(cert);
      new AnyOf([new OIDCIssuer(issuer), new OIDCIssuerV2(issuer)]).verify(
        cert,
      );
      certificateIdentity = identity;
      ref = candidateRef;
    },
  });

  let statementRaw: unknown;
  try {
    statementRaw = JSON.parse(new TextDecoder().decode(verified.payload));
  } catch (cause) {
    throw imageFailure({ digest, reasons: ['invalid_statement'], cause });
  }
  const statement = v.safeParse(ImageProvenanceStatementSchema, statementRaw);
  if (
    !statement.success ||
    verified.payloadType !== 'application/vnd.in-toto+json'
  ) {
    throw imageFailure({ digest, reasons: ['invalid_statement'] });
  }
  const expectedDigest = Buffer.from(digest.slice('sha256:'.length), 'hex');
  const subjectMatches = statement.output.subject.some(
    ({ digest: subjectDigest }) => {
      const sha256 = subjectDigest.sha256;
      return (
        sha256 !== undefined &&
        /^[a-f\d]{64}$/i.test(sha256) &&
        Buffer.from(sha256, 'hex').equals(expectedDigest)
      );
    },
  );
  if (!subjectMatches)
    throw imageFailure({ digest, reasons: ['digest_mismatch'] });

  const commit = verifyImageProvenanceSource({
    statement: statement.output,
    policy,
    ref,
    digest,
  });
  return {
    digest,
    repository: policy.repository,
    workflow: policy.workflow,
    ref,
    commit,
    certificateIdentity,
    issuer,
    predicateType: statement.output.predicateType,
  };
}

type VerifyImageProvenanceSourceParams = {
  statement: ImageProvenanceStatement;
  policy: ImageProvenancePolicy;
  ref: string;
  digest: string;
};

/** Match source fields after the enclosing DSSE statement has been verified. */
export function verifyImageProvenanceSource({
  statement,
  policy,
  ref,
  digest,
}: VerifyImageProvenanceSourceParams): string {
  const repositoryUrl = `https://github.com/${policy.repository}`;
  let commit: string | undefined;
  if (statement.predicateType === 'https://slsa.dev/provenance/v1') {
    const { externalParameters, resolvedDependencies } =
      statement.predicate.buildDefinition;
    const workflow = externalParameters.workflow;
    if (
      workflow.repository.replace(/\.git$/, '') !== repositoryUrl ||
      workflow.path !== policy.workflow ||
      workflow.ref !== ref
    ) {
      throw imageFailure({ digest, reasons: ['source_mismatch'] });
    }
    commit = resolvedDependencies.find(({ uri }) =>
      matchesSourceUri({ uri, repositoryUrl, ref }),
    )?.digest.gitCommit;
  } else {
    const source = statement.predicate.invocation.configSource;
    if (
      !matchesSourceUri({ uri: source.uri, repositoryUrl, ref }) ||
      source.entryPoint !== policy.workflow
    ) {
      throw imageFailure({ digest, reasons: ['source_mismatch'] });
    }
    commit = source.digest.sha1;
  }
  if (commit === undefined || !/^[a-f\d]{40}$/i.test(commit)) {
    throw imageFailure({ digest, reasons: ['source_mismatch'] });
  }
  const normalizedCommit = commit.toLowerCase();
  if (
    policy.commit !== undefined &&
    (!/^[a-f\d]{40}$/i.test(policy.commit) ||
      !Buffer.from(normalizedCommit, 'hex').equals(
        Buffer.from(policy.commit, 'hex'),
      ))
  ) {
    throw imageFailure({ digest, reasons: ['commit_mismatch'] });
  }
  return normalizedCommit;
}

function isSha256Digest(digest: string): boolean {
  return /^sha256:[a-fA-F\d]{64}$/.test(digest);
}

type MatchesSourceUriParams = {
  uri: string;
  repositoryUrl: string;
  ref: string;
};

function matchesSourceUri({
  uri,
  repositoryUrl,
  ref,
}: MatchesSourceUriParams): boolean {
  const source = uri.replace(/^git\+/, '');
  return (
    source === `${repositoryUrl}@${ref}` ||
    source === `${repositoryUrl}.git@${ref}`
  );
}

function invalidApiInput(field: string, expected: string): ApiError {
  return new ApiError({
    code: 'api.invalid_input',
    details: { field, reason: 'unsupported_value', expected },
  });
}

type ImageFailureParams = {
  digest: string;
  reasons: readonly ImageProvenanceFailureReason[];
  cause?: unknown;
};

function imageFailure({
  digest,
  reasons,
  cause,
}: ImageFailureParams): VerificationError {
  return new VerificationError(
    {
      code: 'provenance.image_verification_failed',
      details: { digest, reasons },
      retryable: reasons.includes('trust_root_unavailable'),
    },
    cause === undefined ? undefined : { cause },
  );
}
